### PR TITLE
Provide option to not panic on reconnect retry

### LINKED ISSUE #18: Panic when MaxRetry exceeded
If the fluentd server goes away, panic() is called in a goroutine which causes the entire application to crash. I would like to recover from a panic but seeing as how the panic() is in a separate goroutine that is created by the client library, I can't see how to do that.

Please provide a way to recover from the library panicking.

### PR DESCRIPTION
**Pull request by cpuguy83**
Adds a `Reconnect()` method to force a re-connect.
This can be used once max retries have been hit on attempting to
reconnect to an endpoint.

The caller knows to call `Reconnect()` because posting data will
generate a `ConnectionError` error in the case where the connection is
down.

`Reconnect()` blocks until either max reconnect attemps have been
reached again or a connection is successful.

Fixes #18

### CODE REVIEW

**Comment by cpuguy83** [comment]
ping @tagomoris

**Comment by cpuguy83** [comment]
Ping @tagmoris

**Comment by tagomoris** [comment]
Oops. I'm sorry to miss this so long time.
I'll make sure to review this soon (today, or tomorrow in the worst case).
@cpuguy83 Can you rebase this branch on current master? Or should I do it by myself?

File: fluent/fluent.go [hunk 1]
```diff
@@ -282,31 +289,71 @@ func e(x, y float64) int {
...
+	for f.reconnecting {
+		f.connWait.Wait()
+	}
+
+	if f.conn != nil {
+		return nil
+	}
+
+	f.reconnecting = true
+	go f.reconnect()
```
Comment (tagomoris) [inline fluent/fluent.go]: This looks to be called in parallel with `go f.reconnect()` in L354. Is it correct and intentional?

**Comment by thaJeztah** [comment]
@cpuguy83 looks like this needs a rebase

**Comment by thaJeztah** [comment]
Actually, I see https://github.com/fluent/fluent-logger-golang/issues/18 was closed, and should be fixed by https://github.com/fluent/fluent-logger-golang/pull/56 according to https://github.com/fluent/fluent-logger-golang/issues/18#issuecomment-496765989 so perhaps this one is no longer needed @cpuguy83 ?

**Pull request closed without merging**